core/modules/memman/MemFile.cc

```diff
@@ -46,58 +55,81 @@ std::unordered_map<std::string, MemFile*> fileCache;
 /******************************************************************************/
 
 MemFile::MLResult MemFile::memLock() {
-
+    //
     // The _fileMutex is used here to serialize multiple calls to lock the same
     // file as a file may appear in multiple file sets. This mutex is held for
     // duration of all operations here. It also serialized memory unmapping.
+    // _mlockFileMutex protects _isLocked, allowing _fileMutex to be unlocked
+    // during the mlock call.
     //
-    std::lock_guard<std::mutex> guard(_fileMutex);
-    int rc;
+    std::lock_guard<std::mutex> guardMlock(_mlockFileMutex);
+    int rc = 0;
+    {
+        std::lock_guard<std::mutex> guard(_fileMutex);
+
+        // If the file is already locked, indicate success
+        //
+        if (_isLocked) {
+            MLResult aokResult(_memInfo.size(), _memInfo.mlockTime(), 0);
+            return aokResult;
+        }
```

Comment:
I'm usually skeptical of lock within locks because it may point to a misunderstanding of what the lock is supposed to do and ultimately leads to deadlocks. So, why the change here?

Reply:
The addition of _mlockFileMutex is to reduce the time _fileMutex is held. There are times when the call to memLock can take more than a minute, and having _fileMutex held blocks calls to MemFile::memMap, which causes serious issues with wsched. This is especially the case when a particular file has already been mapped to memory and it is stuck on _fileMutex waiting for an unrelated mlock call.